File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.java
class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        HashMap<String, HashSet<String>> ingredientsMap = new HashMap<>();
        for(int i = 0; i < recipes.length; i++){
            HashSet<String> ingredientsSet = new HashSet<>(ingredients.get(i));
            ingredientsMap.put(recipes[i], ingredientsSet);
        }
        HashSet<String> suppliesSet = new HashSet<>(Arrays.asList(supplies));
        HashMap<String, ArrayList<String>> graph = new HashMap<>();
        for(String u : recipes){
            graph.put(u, new ArrayList<>());
            for(String v : recipes)
                if(ingredientsMap.get(v).contains(u))
                    graph.get(u).add(v);
        }
        ArrayList<String> ordering = topologicalSort(graph);
        ArrayList<String> prepared = new ArrayList<>();
        for(String recipe : ordering){
            if(suppliesSet.containsAll(ingredientsMap.get(recipe))){
                prepared.add(recipe);
                suppliesSet.add(recipe);
            }
        }
        return prepared;
    }
    
    public static <T> ArrayList<T> topologicalSort(HashMap<T, ArrayList<T>> graph){
        HashMap<T, Integer> indegree = new HashMap<>();
        for(T vertex : graph.keySet()) indegree.put(vertex, 0);
        for(T vertex : graph.keySet())
            for(T neighbor : graph.get(vertex))
                indegree.put(neighbor, indegree.get(neighbor)+1);
        Queue<T> queue = new LinkedList<>();
        for(T vertex : indegree.keySet())
            if(indegree.get(vertex) == 0)
                queue.add(vertex);
        ArrayList<T> ordering = new ArrayList<>();
        while(!queue.isEmpty()) {
            T vertex = queue.poll();
            ordering.add(vertex);
            for(T neighbor : graph.get(vertex)) {
                indegree.put(neighbor, indegree.get(neighbor)-1);
                if(indegree.get(neighbor) == 0)
                    queue.add(neighbor);
            }
        }
        return ordering;
        
    }
}

File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.py
from collections import deque
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        
        def topological_sort():
            indegree = {}
            for u in graph:
                if u not in indegree:
                    indegree[u] = 0
                for v in graph[u]:
                    if v not in indegree:
                        indegree[v] = 0
                    indegree[v] += 1
            queue = deque()
            for vertex in graph:
                if indegree[vertex] == 0:
                    queue.append(vertex)
            ordering = []
            while queue:
                vertex = queue.popleft()
                ordering.append(vertex)
                for neighbour in graph[vertex]:
                    indegree[neighbour] -= 1
                    if indegree[neighbour] == 0:
                        queue.append(neighbour)
            return ordering
        
        ingredients = {recipes[i]: set(ingredients[i]) for i in range(len(ingredients))}
        supplies = set(supplies)
        graph = {recipe:[] for recipe in recipes}
        for u in recipes:
            for v in recipes:
                if u in ingredients[v]:
                    graph[u].append(v)
        ordering = topological_sort()
        prepared = []
        for recipe in ordering:
            if ingredients[recipe].issubset(supplies):
                prepared.append(recipe)
                supplies.add(recipe)
        return prepared
    